libraries/protocol/include/graphene/protocol/fba.hpp

```diff
@@ -22,17 +22,17 @@
  * THE SOFTWARE.
  */
 #pragma once
-#include <graphene/chain/protocol/base.hpp>
+#include <graphene/protocol/base.hpp>
 
-namespace graphene { namespace chain {
+namespace graphene { namespace protocol {
 
 struct fba_distribute_operation : public base_operation
 {
    struct fee_parameters_type {};
 
    asset fee;   // always zero
    account_id_type account_id;
-   fba_accumulator_id_type fba_id;
+   object_id_type fba_id;
```

Comment:
This changes the raw serialized representation. It's only half-bad because it's a virtual op, but you still need to bump DB_VERSION because of this.

Reply:
Ahh yeah, this guy... So I can bump the database version, but perhaps at least one comment should be spared on the question: is it appropriate to have an implementation ID type in the protocol library?

Reply:
Hm. I tend to say "no".

In the past we have treated protocol object ids as part of consensus. E. g. we try to not change operation_history IDs even though they never appear on the chain. OTOH we do not guarantee implementation object IDs (e. g. account_history) to remain stable.

`fba_id` has always been an implementation object, thus it should not be used as a persistent identifier but only as a reference within the same database. I think using the generic `object_id_type` here is a good way to communicate this: "Here's an object that will tell you more, but don't make any assumptions about it."

Reply:
Wait, this changes the wire format of an operation... That's not just a DB change, it's a hardfork... And not an easy one unless this operation has never been used before. Or am I mistaken?

It seems that the only realistic choice is to fat-finger define the `fba_accumulator_id_type` in `protocol/types.hpp` with magic numbers that match the ones in `chain/types.hpp`.

I still regard it as inappropriate to reference implementation IDs from operations, but the damage seems fairly permanent now.

Reply:
The better solution is to make the FBA accumulator a protocol type. I'm not well familiarized with FBAs; they were defined while I was away. But having an implementation ID referenced from the protocol, in my opinion, makes our protocol, formally, implementation-defined, and I do not feel that is appropriate.

For better or worse, FBAs are part of the protocol now, and therefore the object ID(s) used to track them should be protocol IDs. Unfortunately, that's still a hardforking change, and a ridiculously difficult one to make. So I'm still stuck at magic numbering an implementation ID into the protocol... This is an architectural travesty.

Reply:
As @pmconrad mentioned it's a virtual operation so didn't be and won't be included in blocks, thus only affects local database.

Reply:
Ack, I completely forgot about that... Still... Yuck. OK.